packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -1,226 +1,148 @@
-  }, [columns]);
-
-  return {
-    columnWidths: columnWidthsRef,
-    setTableWidth,
+export function useTableColumnResizeState<T>(props: TableColumnResizeStateProps<T>, state): TableColumnResizeState {
+  let {
+    getDefaultWidth,
+    getDefaultMinWidth,
+    onColumnResizeStart: propsOnColumnResizeStart,
+    onColumnResizeEnd: propsOnColumnResizeEnd
+  } = props;
+
+  let columnLayout = useMemo(
+    () => new TableColumnLayout({
+      getDefaultWidth,
+      getDefaultMinWidth
+    }),
+    [getDefaultWidth, getDefaultMinWidth]
+  );
+
+  let tableWidth = props.tableWidth ?? 0;
+  let [controlledColumns, uncontrolledColumns]: [Map<Key, GridNode<unknown>>, Map<Key, GridNode<unknown>>] = useMemo(() =>
+    splitColumnsIntoControlledAndUncontrolled(state.collection.columns)
+  , [state.collection.columns]);
+
+  // uncontrolled column widths
+  let [uncontrolledWidths, setUncontrolledWidths] = useState<Map<Key, ColumnSize>>(() =>
+    getInitialUncontrolledWidths(uncontrolledColumns, columnLayout)
+  );
+  // combine columns back into one map that maintains same order as the columns
+  let cWidths = useMemo(() =>
+    recombineColumns(state.collection.columns, uncontrolledWidths, uncontrolledColumns, controlledColumns)
+  , [state.collection.columns, uncontrolledWidths, uncontrolledColumns, controlledColumns]);
+
+  let onColumnResizeStart = useCallback((key: Key) => {
+    columnLayout.setResizingColumn(key);
+    propsOnColumnResizeStart?.(key);
+  }, [columnLayout, propsOnColumnResizeStart]);
+
+  // TODO: move props.on* all into this file and layout, or move them all out to the aria handler..., stately would be preferable
+  let onColumnResize = useCallback((key: Key, width: number): Map<Key, ColumnSize> => {
+    let newControlled = new Map(Array.from(controlledColumns).map(([key, entry]) => [key, entry.props.width]));
+    let newSizes = columnLayout.resizeColumnWidth(tableWidth, state.collection, newControlled, uncontrolledWidths, key, width);
+
+    let map = new Map(Array.from(uncontrolledColumns).map(([key]) => [key, newSizes.get(key)]));
+    map.set(key, width);
+    setUncontrolledWidths(map);
+
+    return newSizes;
+  }, [controlledColumns, uncontrolledColumns, setUncontrolledWidths, tableWidth, columnLayout, state.collection, uncontrolledWidths]);
+
+  let onColumnResizeEnd = useCallback((key: Key) => {
+    columnLayout.setResizingColumn(null);
+    propsOnColumnResizeEnd?.(key);
+  }, [columnLayout, propsOnColumnResizeEnd]);
+
+  // done
+  let getColumnWidth = useCallback((key: Key) =>
+    columnLayout.getColumnWidth(key)
+  , [columnLayout]);
+
+  // done
+  let getColumnMinWidth = useCallback((key: Key) =>
+    columnLayout.getColumnMinWidth(key)
+  , [columnLayout]);
+
+  // done
+  let getColumnMaxWidth = useCallback((key: Key) =>
+    columnLayout.getColumnMaxWidth(key)
+  , [columnLayout]);
+
+  let columnWidths = useMemo(() =>
+      columnLayout.buildColumnWidths(tableWidth, state.collection, cWidths)
```

Comment:
This line confused me a bit since `buildColumnWidths` takes `controlledWidths` as its third arg but cWidths here is just all the column widths (controlled and uncontrolled). Mind renaming the third input arg to `buildColumnWidths` as `calculatedColumnWidths` or something instead?

Reply:
that's fair